.github/workflows/ci.yaml

```diff
@@ -0,0 +1,137 @@
+name: CI
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+env:
+  CARGO_TERM_COLOR: always
+  RUST_BACKTRACE: full
+jobs:
+  pre-commit:
+    name: Validate `pre-commit` Hooks
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout sources
+      uses: actions/checkout@v3
+    - name: Get python
+      uses: actions/setup-python@v3
+    - name: Run pre-commit check
+      uses: pre-commit/action@v3.0.1
+      with:
+        extra_args: run --all-files --hook-stage manual
+  conventional-commits-check:
+    name: Conventional Commit Checking
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@main
+        with:
+          fetch-depth: 0
+      - name: Assert conventional commit usage
+        uses: oknozor/cocogitto-action@v3
+  spell-check:
+    name: Spell Checking
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout sources
+      uses: actions/checkout@v4
+    - name: typos
+      uses: crate-ci/typos@master
+  broken-url-check:
+    name: Broken URL Checking
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v4
+      - name: Run lychee
+        uses: lycheeverse/lychee-action@v1.9.3
+        with:
+          fail: true
+  toml-check:
+    name: TOML Validation
+    runs-on: ubuntu-latest
+    container:
+      image: tamasfe/taplo:0.8.0
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v4
+      - name: taplo lint
+        run: taplo lint
+      - name: taplo fmt
+        run: taplo fmt --check --diff
+  rust-format-check:
+    name: Rust Code Format Checking
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v4
+      - name: Build setup
+        uses: ./.github/actions/build-setup
+      - name: cargo fmt
+        run: cargo fmt --all -- --check
+  rust-lint-check:
+    name: Lint Checking
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v4
+      - name: Build setup
+        uses: ./.github/actions/build-setup
+      - name: Run clippy
+        run: cargo clippy --all-features --tests -- -D warnings
+  rust-docs-check:
+    name: Developer Documentation Validation
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout sources
+        uses: actions/checkout@v4
+      - name: Build setup
+        uses: ./.github/actions/build-setup
+      - name: Run cargo doc
+        env:
+          RUSTDOCFLAGS: "-D warnings"
+        run: cargo doc --no-deps --all-features
```

Comment:
suggestion: I'd rather keep the actions language-agnostic and move in the TOML-stuff in e.g. the lint-step and the format-step.

Reply:
Hmm. I'm personally of the opinion that a job should do one thing and one thing only. It should be possible to simply glance at which jobs fail and know what actions to take to fix them, without necessarily needing to go dig in the job logs. Running too many commands for different "languages"  inside a single job goes against this, whilst also making it harder to separate their context in case of required dependencies, artifact, and cache collisions. I'm ok with splitting the TOML lint and format steps into two separate jobs. I usually try to balance runner compute requirements with the above reasoning.

Reply:
Yeah, I too find the idea of not exposing the inner tool more than necessary :)